### ISSUE TITLE
Consistency checking in mods?

### ISSUE THREAD

**Issue by johnstt** [ValveSoftware/halflife#1596]
Hi! 
I would like to know how I can add consistency checking for arbitrary files in my mod. I'm looking for what triggers the message "Your file ... differs from the server's" and forces clients to disconnect.

More specifically, I want to make sure the client/server DLLs (and .so binary) are checked. I am updating a mod that was last released 9 years ago, and I want to make the process of beta testing with the game's community as ..consistent.. as I can.
Is pfnForceUnmodified the correct call to use? I tried this but it did not seem to work. Any advice?

**Comment by JoelTroch** [comment]
I think that the consistency check for binaries is done on the engine. Have you tried to set up a dedicated server with the "good" binaries and simulate a client using "bad" binaries ?

**Comment by johnstt** [comment]
Yes, I can join a server that is running the new binaries while I have the old ones (and vice versa).

**Comment by JoelTroch** [comment]
Even with "sv_pure" enabled ?

**Comment by johnstt** [comment]
It looks like I don't 'have' sv_pure.
Nothing about it in source code, doesn't give a value through rcon, setting it doesn't seem to do anything...

**Comment by JoelTroch** [comment]
"mp_consistency 1" ?

**Comment by johnstt** [comment]
mp_consistency uses pfnForceUnmodified in our source code.
So, the first test I mentioned was me extending that to apply to the DLLs. Didn't work.

**Comment by JoelTroch** [comment]
I think this is a "hardcore" solution but worth a try : make a CVAR which can't be modified which contains your mod version and check somewhere (like player connect) to see if they match.

**Comment by sylar0214** [comment]
there is also a plugin filescheck.amxx (i have it, cant find the link) in which you can add your sounds, sprites etc and can check if client is missing or if it is different.

**Comment by johnstt** [comment]
How does the plugin implement it?
That's what I want to know, so I don't have to add a plugin to my game to get something that's already built-in.

**Comment by GuyonthecoucH** [comment]
To clarify Tim's question we are primarily concerned with checking if someone is running the most up to date version of the mod. Our current situation is like being able to join a Counter-Strike 1.6 server with with the Counter-Strike 1.5 client.

If I recall correctly if a user tried to join a server with an outdated client, they wouldn't connect and would get a pop-up message stating something like "your version differs from the server" or something. That's what we are looking for.

**Comment by tschumann** [comment]
So the engine didn't complain when using a different client.dll?

**Comment by johnstt** [comment]
Right. But I know half-life *can* complain about that...I'm missing a server setting or something in the game code needs to be changed.

**Comment by tschumann** [comment]
What does your liblist.gam look like?

**Comment by johnstt** [comment]
Here's the one from the last release in 2006. We're modifying it for the next release by adding edicts "4096" to address an unrelated issue...

```
// Liblist.gam mod setup file for Science and Industry
// Load order is determined by file order.

game "Science and Industry"
url_info "http://si.planethalflife.gamespy.com/"
url_dl "http://si.planethalflife.gamespy.com/downloads.php"
manual "http://si.planethalflife.gamespy.com/manual.php"
version "1.1"
size "210000000"
svonly "0"
cldll "1"
hlversion "1108"
type "multiplayer_only"
nomodels "1"
mpentity "info_sidetect"
gamedll "dlls\si.dll"
gamedll_linux "dlls/si_i386.so"
icon "resource\si_icon"
```

**Comment by FrenchMan** [comment]
@johnstt 
Try adding the following line to the liblist.gam:
```
crcclientdll "1"
```
It will notify the game to check the client library.

**Comment by JoelTroch** [comment]
First time I see ```crcclientdll "1"``` to be added on a "liblist.gam" file, what about ```secure "1"``` ?

**Comment by FrenchMan** [comment]
AFAIK `secure 1` means VAC protection, not the resources consistency checks.

**Comment by johnstt** [comment]
Thanks, guys. I'll try these options.

*Just to clarify:* the liblist.gam handling this is serverside. I take it nobody is handling this with game code or svars? It seems like either of those would be more natural...

**Comment by FrenchMan** [comment]
@johnstt 
I hope it is server-side, it should be at least. :) The `crcclientdll` option tells engine to inform a player if the client library differs from the server's. Literally, the engine compares checksums. So to theck if this works, cl and sv should have different client dll/so/dylib.

**Comment by johnstt** [comment]
How do I know if my server is doing the check?
(I ask because I added the option to my liblist and still connected with several different client DLLs)

**Comment by FrenchMan** [comment]
@johnstt 
Hmm, that's interesting. I can only suppose that the engine uses checksum from the library header, and you've builded mod libraries without filling the checksum (leaving it zero is the default behavior for debug-mode libs in MSVS). Anyway, it's hard to say what the engine exactly does, maybe @alfred-valve will come and answer? :)

**Comment by ripieces** [comment]
Don't get your hopes too high, I reported a SSE2 instruction set usuage regression bug in i.e. Steam client's v8.dll on 29th January 2015 [here](https://support.steampowered.com/view.php?ticketref=2524-OJVB-8385) and so far nothing happened (neither fixed nor system requirements cranked up).
I am glad that I have access to a newer laptop, with a CPU that supports SSE2, but if that wouldn't be the case, I'd be really sad.

I don't know with what Valve is busy since September 2014, but it must consume quite a lot of their time.

**Comment by ripieces** [comment]
What I meant to say is you are better of implementing an own check that is communitcated between client.dll and server.dll than waiting for them. Of course such checks can be circumvented, but still better than nothing.

If you need s.th. simple, you could try if
ENGINE_FORCE_UNMODIFIED aka [pfnForceUnmodified](https://github.com/ValveSoftware/halflife/blob/5c1e1d63eb5a9cb97b3b19f992d4e62aaae02cf0/engine/eiface.h#L258) works for abitrary files in force_exactfile mode.
An example usage is in DMC mod's client.dll in ClientPrecache [here](https://github.com/ValveSoftware/halflife/blob/5c1e1d63eb5a9cb97b3b19f992d4e62aaae02cf0/dmc/dlls/client.cpp#L807).
it will trigger InconsistentFile function in client.dll [here](https://github.com/ValveSoftware/halflife/blob/5c1e1d63eb5a9cb97b3b19f992d4e62aaae02cf0/dmc/dlls/client.cpp#L1789).

I don't know how efficient that is coded on the server side, so you might be better of implementing your own check.

**Comment by johnstt** [comment]
Of course I could implement something myself.
However, reinventing the wheel is generally a bad idea for a variety of reasons. A forum like this should be a place where I can get closer to an answer. 

I'm almost totally certain this is a standard feature of the game, and I just don't have the right settings to use it (yet). Not something for individual mod programmers to solve.

In the case of the pfnForceUnmodified engine function, I mentioned right at the start that I tried it and it didn't work.

**Comment by FrenchMan** [comment]
@johnstt 
Oh, I forget to mention, are you using the last beta?

**Comment by johnstt** [comment]
Beta of what?

**Comment by FrenchMan** [comment]
Beta of game/engine.

**Comment by johnstt** [comment]
Hmm that thread is a bit depressing!
So if I need to implement it myself, maybe the next thing is to see why pfnForceUnmodified doesn't work, since that's (ideally) the way to do it. Since the liblist.gam option doesn't work for me, I would guess it's due to (possibly huge) differences between the SDK code and my mod's code from 2006. I don't want to even open that can of worms right now...

As for the beta:
I want this to apply to clients of both (since I can't control what people use to play), and be applied *by the server* somehow. I guess I'll poke at it myself unless someone shows up here with a magic solution.

Thanks, guys :(

**Comment by ripieces** [comment]
@johnstt This works only for precached files, you can use PRECACHE_GENERIC to precache generic files!

So you would do this in ClientPrecache in dlls\client.dll
PRECACHE_GENERIC("cl_dlls/client.dll")
ENGINE_FORCE_UNMODIFIED(force_exactfile, Vector(0,0,0), Vector(0,0,0),"cl_dlls/client.dll")

I tried it and it works this way.
Not sure if it's the right way to do it, since I don't know if it's a waste of memory to precache it.

Btw in NaturalSelection mod they made a mistake, the use it to check sounds, but apparently forgot to append the sound/ folder prefix ;) (Meaning it won't work for those files).

**Comment by johnstt** [comment]
Aha! Very cool. I'll try this soon, thanks!

**Comment by Freeman-AM** [comment]
Can this block some cheat ?
If so, i would love to have a module version of this.

**Comment by JoelTroch** [comment]
@Freeman-AM I don't think it will be block cheats. I may be wrong but deserves a try.

@ripieces To be honest, I prefer to write a custom checksum check than wasting memory with a  ```PRECACHE_GENERIC```, but if it works that way, it's good.
Is it necessary to precache the DLL ? You can't do without it ?

**Comment by ripieces** [comment]
@Freeman-AM This won't block most cheats, because those modify the loaded DLL file  in memory and not the file on disk from where it is precached and compared.

@JoelTroch I sadly don't know the hw.dll source code, so I can't say for sure, but it makes sense that it works on precached files and doesn't seem to work on non-precached files. If you look at any sample code, i.e. of AMX Mod X, then the file(s) are always precached before the ENGINE_FORCE_UNMODIFIED is used.

**Comment by ripieces** [comment]
@Johnstt Just so you guys know, you can only PRECACHE_GENERIC for 512 (0x200) items!
So avoid checking more generic files than necessary.

Same goes for PRECACHE_SOUND and PRECACHE_MODEL, they have a limit of 512 (0x200) items **each**.

Sinc I doubt you are using PRECACHE_GENERIC in your mod for anything else than client.dll at the moment, you should be fine with that limit though I guess.

**Comment by ripieces** [comment]
And another thing to note:
The file will be downloaded to  the client in case it doesn't exist it seems.
But I assume your mod will always have a client.dll, so that shouldn't be a problem.

Maybe it's really better if you code a DLL version / checksum check yourself somehow :-(

**Comment by johnstt** [comment]
I still have yet to sit down and write my own DLL checker, but a thought occurs to me:
Isn't consistency still checked for maps? If it is, then I guess the next question is obvious...

**Comment by ripieces** [comment]
It's not obvious to me?
I don't know if they go through PRECACHE_GENERIC or not, I just remember that they are allowed to differ in dynamic entities between client and server (those will be spawned on the client from the server).

**Comment by tschumann** [comment]
Here's the reason why this used to work out of the box but now doesn't: https://github.com/ValveSoftware/halflife/issues/564

**Comment by johnstt** [comment]
Yeah, and I really disagree that it's "useless for games on steam and gets in the way".

My mod has been dealing with versioning issues for the past year that would have been solved for free a few years ago (or if the code for it simply **was not removed**). 

Most mods don't have steampipe access, so they need this feature. When I sit down to work on my mod, I don't want to reinvent the wheel (especially when I'm limited in what I can do by the engine and hlds).
